### ISSUE TITLE
'Block Input to Background Apps' not blocking input to games

### ISSUE THREAD

**Issue by CheesesteakVR** [Xiexe/XSOverlay-Issue-Tracker#273]
**Describe the bug in detail:**
'Block Input to Background Apps' should block input to running games, but does not. This is after ensuring 'Enable global input from overlays (Experimental)' is enabled in Steam developer settings, and that all other SteamVR overlay applications are disabled.

Weirdly enough, SOME inputs are blocked; turning left and right in games (right hand joystick left and right) are properly blocked, for instance. (This is the only known exception.)

Issue persists between multiple apps, e.g. Skyrim VR and VRChat.

Issue persists between multiple XSOverlay versions, namely stable and beta - beta (build 614 UI3.0 RC2).

I _strongly_ suspect this is related to WMR and its interaction with SteamVR.

Video: https://cdn.discordapp.com/attachments/628376970203496448/1148115464782876752/20230904_003138_MixedReality-00.00.00.000-00.00.33.635.mp4

**Provide steps to reproduce the bug:**
- In XSOverlay settings, enable 'Block Input to Background Apps'
- Toggle layout mode
- Interact with any UI in XSOverlay

**Expected behavior:**
I expect inputs such as trigger on each controller to be blocked, so that one does not e.g. accidentally press buttons or pick objects up in a game like VRChat. Instead, inputs are propagated through to the running app.

**Output Log:**
[output_log_2023-09-04__12.15.19.txt](https://github.com/Xiexe/XSOverlay-Issue-Tracker/files/12515577/output_log_2023-09-04__12.15.19.txt)
[output_log_2023-09-04__12.16.37.txt](https://github.com/Xiexe/XSOverlay-Issue-Tracker/files/12515578/output_log_2023-09-04__12.16.37.txt)
(uWindowCapture and uDesktopCapture log in the XSOverlay install directory not present)

**Additional Information:**
Windows 11 Pro 22H2
SteamVR Beta 1.27.5
Windows Mixed Reality for SteamVR [Beta]

**Comment by Xiexe** [comment]
> I strongly suspect this is related to WMR.

Yeah, I'm not able to recreate this on either of the platforms I have for testing (Index and Oculus)

SteamVR only blocks inputs that it sees as taken by the currently loaded binding, and anything else is ignored.

It seems likely that they've inverted some logic somewhere internally for WMR bindings, since you said it's blocking left and right turning. That's pretty much the only input XSO doesn't use.

**Comment by CheesesteakVR** [comment]
Apologies, I totally forgot to follow up with further investigation I've done on this issue.

It turns out that yes, it does properly block inputs based on bindings, and I wasn't being rigorous enough. Grip is bound in XSOverlay, and that is properly blocked. The touchpads are bound, and that is properly blocked. The right stick is by default bound to desktop scroll while the left stick is unbound, so binding something like desktop scroll to the left stick also properly blocks the left stick (This is why I had reported it as "Turning is blocked".)

In summary, every input appears to be properly blocked when told to, EXCEPT for the triggers.

![20231017_152513_MixedReality](https://github.com/Xiexe/XSOverlay-Issue-Tracker/assets/143066815/e8c184f2-3900-408f-aecf-d58aa7f05b3c)
![20231017_152530_MixedReality](https://github.com/Xiexe/XSOverlay-Issue-Tracker/assets/143066815/7b041c08-0751-4ee2-b70e-cf2d9752a617)
These are the trigger bindings I have for XSOverlay and VRChat (This also persists in other applications, so it's not a VRChat exclusive issue; I use Trigger Value (Or the closest thing to it I can) for those bindings.

I've spoken with another WMR user, and they have repro on this bug.

The screenshots are in the current SteamVR Beta, although this behavior is the same for stable.

**Comment by Xiexe** [comment]
If everything except trigger pulls is being blocked, that sounds like a SteamVR bug.

**Comment by CheesesteakVR** [comment]
I think I've identified a possible workaround/fix for this issue.

1. Go into the SteamVR overlay to bindings.
2. For the triggers on each controller, change them from 'trigger' to 'button'.
3. Bind an action to them. (I used Left Click.)
4. Apply.
5. Change it back from 'button' to 'trigger'.
6. Rebind them to trigger value.
7. Apply.

This will make the input actually be blocked properly. I've verified that this works for SteamVR Stable too. 

I'm bad at explaining via text, so here's a short video clip.
https://youtu.be/c6qxsQdSOxk

I speculate that behind the scenes, the bind for the button is being saved into a file somewhere, and then being disused on the actual controller but being read as bound for the purposes of input blocking. So it now knows to block the input thanks to the dummy binding. If this is the case, perhaps a config like this can be provided by default for this platform?

**Comment by Xiexe** [comment]
If you bind the button on the trigger to also be trigger value, does that work?

On index controllers you can bind trigger value to a button and it'll just send a float value of 1, so the pull action will continue to work and it'll just send 1 when the trigger is fully depressed, which it would do anyways.

That may be a quick fix for this without affecting functionality.

If that does work, I will update the default bindings files to reflect that.

If not, I can add a new binding called "WMR Dummy Bind" or something of that nature.

**Comment by CheesesteakVR** [comment]
Leaving the bind on Use as Button -> Trigger Value is not sufficient. it needs to be bound to Use as Button, and then changed back to Use as Trigger -> Trigger Value.
![20231026_180440_MixedReality](https://github.com/Xiexe/XSOverlay-Issue-Tracker/assets/143066815/17fcb107-50e7-4f84-aa4a-691a733b4908)
Attachment: not enough to block input on trigger.

**Comment by Xiexe** [comment]
Do the bindings not let you set both?

Pull: Trigger Value
Click: Trigger Value

If not I may be misremembering how the bindings page works.

**Comment by CheesesteakVR** [comment]
I can bind "Use as Button" and "Use as Trigger" at the same time if that's what you mean. I've tried a few permutations of the different combinations of those two and other binding options at the same time to try and get this blocking to work, and the only way I've been able to get it to work is through the method outlined above. Once that's saved, I can change configs back to default and then load the "fixed" config and it works properly. I've published it as "Default Binds W/ Input Block Workaround" on Steam if you want to examine it.
